tee-worker/omni-executor/rpc-server/src/methods/omni/submit_user_op.rs

```diff
@@ -0,0 +1,192 @@
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with Litentry.  If not, see <https://www.gnu.org/licenses/>.
+
+use super::common::handle_omni_native_task;
+use crate::error_code::AUTH_VERIFICATION_FAILED_CODE;
+use crate::methods::omni::common::check_auth;
+use crate::methods::omni::PumpxRpcError;
+use crate::server::RpcContext;
+use crate::ErrorCode;
+use alloy::primitives::{Address, FixedBytes};
+use executor_core::native_task::{NativeTask, NativeTaskWrapper};
+use executor_core::types::SerializablePackedUserOperation;
+use executor_primitives::{AccountId, ChainId};
+use jsonrpsee::RpcModule;
+use native_task_handler::NativeTaskOk;
+use parity_scale_codec::Decode;
+use serde::{Deserialize, Serialize};
+use std::collections::HashSet;
+use tracing::{debug, error};
+
+#[derive(Debug, Deserialize)]
+pub struct SubmitUserOpParams {
+	pub user_operations: Vec<SerializablePackedUserOperation>,
+	pub chain_id: ChainId,
+	pub wallet_index: u32,
+}
+
+#[derive(Serialize, Clone)]
+pub struct SubmitUserOpResponse {
+	pub transaction_hash: Option<String>,
+}
+
+pub fn register_submit_user_op(module: &mut RpcModule<RpcContext>) {
+	module
+		.register_async_method("omni_submitUserOp", |params, ctx, ext| async move {
+			let user = check_auth(&ext).map_err(|e| {
+				error!("Authentication check failed: {:?}", e);
+				PumpxRpcError::from_error_code(ErrorCode::ServerError(
+					AUTH_VERIFICATION_FAILED_CODE,
+				))
+			})?;
+
+			let params = params.parse::<SubmitUserOpParams>().map_err(|e| {
+				error!("Failed to parse params: {:?}", e);
+				PumpxRpcError::from_error_code(ErrorCode::ParseError)
+			})?;
+
+			debug!("Received omni_submitUserOp, params: {:?}", params);
+
+			let address_bytes =
+				hex::decode(user.omni_account.strip_prefix("0x").unwrap_or(&user.omni_account))
+					.map_err(|_| {
+						error!("Failed to decode omni account hex string");
+						PumpxRpcError::from_error_code(ErrorCode::InternalError)
+					})?;
+
+			if address_bytes.len() != 32 {
+				error!(
+					"Invalid omni account length: expected 32 bytes, got {}",
+					address_bytes.len()
+				);
+				return Err(PumpxRpcError::from_error_code(ErrorCode::InternalError));
+			}
+
+			let mut address = [0u8; 32];
+			address.copy_from_slice(&address_bytes);
+
+			// Collect unique sender addresses to avoid redundant validation calls
+			let unique_addresses: HashSet<Address> = params
+				.user_operations
+				.iter()
+				.map(|op| {
+					op.sender.parse::<Address>().map_err(|e| {
+						error!("Invalid sender address '{}': {}", op.sender, e);
+						PumpxRpcError::from_error_code(ErrorCode::ParseError)
+					})
+				})
+				.collect::<Result<HashSet<_>, _>>()?;
+
+			// Validate each unique address once
+			for sender_address in unique_addresses {
+				if let Err(e) = validate_user_operation_ownership(
+					&user,
```

Comment:
How about if we pass `address_bytes` here so we don't have to extracted again from `user`?

Reply:
resolved